cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -583,17 +614,25 @@ public Void call() throws Exception {
               }
             }
             context.setState(getProgramState(success, failureInfo));
-            if (mapReduce instanceof ProgramLifecycle) {
-              ((ProgramLifecycle) mapReduce).destroy();
-            } else {
-              mapReduce.onFinish(success, context);
+            if (TransactionControl.IMPLICIT == txControl) {
+              doDestroy(success);
             }
-            return null;
+            return success;
           } finally {
             ClassLoaders.setContextClassLoader(oldClassLoader);
           }
         }
       });
+      if (TransactionControl.EXPLICIT == txControl) {
+        ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(mapReduceClassLoader);
```

Comment:
Better to move the setting and resetting of classloader inside the `doDestroy` method (similarly for `doInitialize`) to reduce code duplication.